internal/infrastructure/translator.go

```diff
@@ -0,0 +1,56 @@
+package infrastructure
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	clicfg "sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	"github.com/envoyproxy/gateway/api/config/v1alpha1"
+	"github.com/envoyproxy/gateway/internal/envoygateway/config"
+	"github.com/envoyproxy/gateway/internal/infrastructure/kubernetes"
+	"github.com/envoyproxy/gateway/internal/ir"
+)
+
+// Translate translates the provided infra ir into managed infrastructure.
+func Translate(ctx context.Context, cfg *config.Server, infra *ir.Infra) (*Context, error) {
+	if err := ir.ValidateInfra(infra); err != nil {
+		return nil, err
+	}
+	if cfg == nil {
+		return nil, errors.New("server config is nil")
+	}
+
+	infraCtx, err := NewContext(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	log := cfg.Logger
+
+	if *infraCtx.Provider == v1alpha1.ProviderTypeKubernetes {
+		log.Info("Using infra manager", "type", v1alpha1.ProviderTypeKubernetes)
+		cli, err := client.New(clicfg.GetConfigOrDie(), client.Options{})
+		if err != nil {
+			return nil, err
+		}
+		kubeCtx := kubernetes.NewContext(cli, cfg)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create kube infra manager context: %v", err)
+		}
+		infraCtx.Kubernetes = kubeCtx
+
+		// A nil infra proxy ir means the proxy infra should be deleted, but metadata is
+		// required to know the ns/name of the resources to delete. Add support for deleting
+		// the infra when https://github.com/envoyproxy/gateway/issues/173 is resolved.
```

Comment:
imho the metadata should only be used for informational features such as labelling and status and not for CRUD for k8s resources.
one way to handle deletes would be for the infra manager to reconcile all k8s resources owned by this infra manager (tied to a single GW Class), and delete those resources that are not part of the Translate o/p

Reply:
IMO infra IR metadata should include a `deletionTimestamp` that is set by the Resource Translator when the associated Gateway is deleted. An infra IR with `deletionTimestamp` tells the Infra Manager to delete the associated infra.

xref kube meta: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata

> one way to handle deletes could be for the infra manager to reconcile all k8s resources owned by this infra manager (tied to a single GW Class), and delete those resources that are not part of the Translate o/p

Yes, this is a possible solution but IMO not as efficient as it requires a List() and then Delete().